### ISSUE TITLE
StaticDecode incorrectly handles Optional properties.

### ISSUE THREAD

**Issue by HamishWHC** [sinclairzx81/typebox#543]
Hi! I'm having an issue with transform types (and thus StaticDecode, to get a TS type for the decoded structure) and optionals. I'm seeing the following problem:

```ts
const ExampleOptional = Type.Object({
	string: Type.Optional(Type.String()),
	object: Type.Optional(Type.Object({})),
	union: Type.Optional(Type.Union([Type.String(), Type.Number()])),
});

type ExampleOptionalDecoded = StaticDecode<typeof ExampleOptional>;
/*
{
    string?: string | undefined;
    object: {};
    union: string | number;
}
*/

// The encoded version handles them correctly.
type ExampleOptionalEncoded = StaticEncode<typeof ExampleOptional>;
/*
{
    string?: string | undefined;
    object?: {} | undefined;
    union?: string | number | undefined;
}
*/

const value: unknown = {};
if (Value.Check(ExampleOptional, value)) {
	const decoded: ExampleOptionalDecoded = Value.Decode(ExampleOptional, value);
	console.log("decoded value", decoded); // logs `{}`, despite being typed as having two required properties
}
```

This isn't a problem in the above schemas since they don't need decoding, but I want to use optionals in a schema with Transform types, so this is causing problems.

Thanks for your help.

**Comment by HamishWHC** [comment]
Did some poking around and found that there is no case for TOptional in DecodeStaticType, which resolves the example given. I was able to patch it with the following:
```ts
T extends TOptional<infer S> ? TOptional<DecodeStaticType<S>>
```
ExampleOptionalDecoded and Encoded now match.

However, when adding a transform to the object:
```
	transform: Type.Optional(
		Type.Transform(Type.Array(Type.String()))
			.Decode((value) => new Set(...value))
			.Encode((value) => [...value]),
	),
```
The ExampleOptionalDecoded type is now 
```ts
{
    string?: string | undefined;
    object?: {} | undefined;
    union?: string | number | undefined;
    transform: Set<string>;
}
```

This seems to be because TOptional marks the property as optional by adding a symbol to the interior type, but this type is TUnsafe since it is a decoded transform type. I was able to patch this with `T extends TOptional<TTransform<infer _, infer R>> ? TUnion<[TUnsafe<R>, TUndefined]>` added to the beginning of DecodeStaticType but it feels rather hacky due to the nested infer and this only gets `| undefined` rather than making the property optional: ```ts
{
    string?: string | undefined;
    object?: {} | undefined;
    union?: string | number | undefined;
    transform: Set<string> | undefined;
}
```

I'll keep playing around with it :)

**Comment by HamishWHC** [comment]
Figured it out! TOptional was being removed by any of the conditions, since (for example) `TOptional<TUnion<infer S>>` *also* extends `TUnion<infer S>`. This would then return `TUnion<AssertRest<DecodeStaticRest<S>>>`, which has no mention of TOptional.

To fix, I replaced my previous patches with a single TOptional case at the beginning: `T extends TOptional<infer S> ? TOptional<DecodeStaticType<S>>` which will unwrap optionals before other schemas. I'll create a PR with this now.

**Comment by sinclairzx81** [comment]
@HamishWHC Hi!

Hey, thanks for the PR, but just looking into this myself (as it's a bit of critical one this one given the potential to exceed recursive instantiation limits). 

Needs to be handling both Readonly and Optional (so it's a bit of a tricky one), but somewhat curious as to why TS is unable to resolve for the modifiers in the `TObject` case only.

Currently have the following setup for `DecodeStaticProperties`

```typescript
export type DecodeStaticProperties<T extends TProperties> = {
  [K in keyof T]: 
    T[K] extends TReadonly<T[K]> & TOptional<T[K]> ? TReadonly<TOptional<DecodeStaticType<T[K]>> : 
    T[K] extends TReadonly<T[K]> ? TReadonly<DecodeStaticType<T[K]>> :
    T[K] extends TOptional<T[K]> ? TOptional<DecodeStaticType<T[K]>> :
    DecodeStaticType<T[K]>
}
```
Just investigating.

**Comment by HamishWHC** [comment]
Ah I see, yeah I hadn't considered Readonly since I'm not using it myself. I was initially playing around with PropertyReducer to try and reuse that logic here, but wasn't able to get it working.

**Comment by sinclairzx81** [comment]
@HamishWHC Hey thanks for taking a look! (and thanks for reporting too!)

I think you were on the right track with mapping within the `StaticDecodeType` as this type is naively reassigning the container type (TObject, TUnion, TArray, etc, etc) without considering the modifiers applied to the type (so totally a bug)

I think to solve this in a general way a new `StaticDecodeModifier<T extends TSchema, Decoded>` may be required which is used on the Decoded container types.

**Comment by HamishWHC** [comment]
Are these modifiers limited by where they can appear (i.e. only on object properties)? If not, it should be possible to simply update DecodeStaticType to check for modifiers first like so, which avoids needing to replace existing references to it:

```ts
export type DecodeStaticType<T extends TSchema> =
  T extends TReadonly<infer S> ? TReadonly<DecodeStaticType<S>> :
  T extends TOptional<infer S> ? TOptional<DecodeStaticType<S>> :
  // ... remainder of DecodeStaticType
```

The modifiers will always be handled first and the inferred type S doesn't contain the modifier symbol, so this will not recurse infinitely as far as I can tell. Testing this out, it seems to work fine:

```ts
const ROPerms = <T extends TSchema>(schema: T) =>
	Type.Object({
		neither: schema,
		optional: Type.Optional(schema),
		readonly: Type.Readonly(schema),
		readonly_optional: Type.Readonly(Type.Optional(schema)),
		optional_readonly: Type.Optional(Type.Readonly(schema)),
	});

const Example = Type.Object({
	string: ROPerms(Type.String()),
	union: ROPerms(Type.Union([Type.String(), Type.Number()])),
	object: ROPerms(Type.Object({ test: Type.String() })),
	transform: ROPerms(
		Type.Transform(Type.Array(Type.String()))
			.Decode((value) => new Set(...value))
			.Encode((value) => [...value]),
	),
});

type ExampleDecoded = StaticDecode<typeof Example>;
/*
{
    string: {
        readonly readonly_optional?: string | undefined;
        readonly optional_readonly?: string | undefined;
        readonly readonly: string;
        optional?: string | undefined;
        neither: string;
    };
    object: {
        readonly readonly_optional?: {
            test: string;
        } | undefined;
        readonly optional_readonly?: {
            test: string;
        } | undefined;
        readonly readonly: {
            test: string;
        };
        optional?: {
            test: string;
        } | undefined;
        neither: {
            test: string;
        };
    };
    union: {
        readonly readonly_optional?: string | number | undefined;
        readonly optional_readonly?: string | number | undefined;
        readonly readonly: string | number;
        optional?: string | number | undefined;
        neither: string | number;
    };
    transform: {
        readonly readonly_optional?: Set<string> | undefined;
        readonly optional_readonly?: Set<string> | undefined;
        readonly readonly: Set<string>;
        optional?: Set<string> | undefined;
        neither: Set<string>;
    };
}
*/
```

**Comment by HamishWHC** [comment]
Have just realised I missed the comments in the PR, sorry!

**Comment by sinclairzx81** [comment]
@HamishWHC Hey, have just pushed an update to your PR branch. 

You were 100% right that the mapping should occur within `StaticDecodeType` :) Looking deeper, I see the issue was that the container types (i.e. `TObject`, `TUnion`, `TIntersect`) were just natively remapping without re-assignment of the modifier (which is a bit of an oversight on my part)

The patch I've pushed to your branch remaps explicitly for each container type via a new `DecodeModifier` type. I've made a small change there to remove the `TOptional` check from the `StaticDecodeType` specifically, and moved it into it's own type named `DecodeModifier<D, T>`. The reason for this is that the `StaticDecodeType` is used as a visitor type (i.e. type traversal) and that TB treats modifiers as non-traversable metadata via `T & modifier`

Object Decode now looks as follows.

```typescript
export type DecodeModifier<D extends TSchema, T extends TSchema> =
  T extends TReadonly<T> & TOptional<T> ? TReadonly<TOptional<D>> :
  T extends TReadonly<T> ? TReadonly<D> :
  T extends TOptional<T> ? TOptional<D> :
  D

...

T extends TObject<infer S> ? DecodeModifier<TObject<Evaluate<DecodeProperties<S>>>, T>
                          // ^ yikes - but remaps the modifier (post) decode
```
This should hopefully resolve the issue at minimal additional type instantiations during decoding (these Decode types are super expensive)

Will look to merging the branch through within the hour.
Cheers
S
